### PR TITLE
feat: Add TokenUpdateNFTs as a smart contract operation

### LINKED ISSUE #14589: Define and implement HTS function call TokenUpdateNfts

### LINKED ISSUE #14620: Write E2E tests for system contract enhancement for TokenUpdateNfts

### PR DESCRIPTION
**Pull request by MrValioBg**
**Description**:
- Add `TokenUpdateNFTs` as a smart contract operation - `updateNFTsMetadata(address, int64[], bytes)`
- Add numeric validation test for the serial numbers of the new smart contract operation.
- Create `UpdateTokenMetadataTest.java` and add success cases. Since we want to add new metadata-related operations soon, this test class can be used for them as well.
- Add unit tests for new classes.

**Related issue(s)**:

Fixes #14589 #14620 

**Notes for reviewer**:

**TODO**

- [ ] After confirmation of all the changes in this PR, I will open one in smart-contracts repo, to apply the changes to `IHederaTokenService.sol` and `HederaTokenService.sol`

### CODE REVIEW

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/SharedNetworkLauncherSessionListener.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
  * plan execution finishes.
  */
 public class SharedNetworkLauncherSessionListener implements LauncherSessionListener {
-    public static final int CLASSIC_HAPI_TEST_NETWORK_SIZE = 4;
+    public static final int CLASSIC_HAPI_TEST_NETWORK_SIZE = 1;
```
[the remark this answers is not in the document]
Comment (MrValioBg) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/SharedNetworkLauncherSessionListener.java]: Yup, thanks. I've removed it.

**Comment by MrValioBg** [comment]
> We can add negative cases like calling update without having the metadata key

@mustafauzunn 
I haven't added such a tests, as we already test most of the negative test scenarios in TokenUpdateNftsSuite, including the one without a metadata key.

The tests included in this PR, are focused mainly on testing the smart-contract interactions.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/update/UpdateNFTsMetadataTranslator.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import javax.inject.Inject;
+
+public class UpdateNFTsMetadataTranslator extends AbstractCallTranslator<HtsCallAttempt> {
+    public static final Function UPDATE_NFTs_METADATA =
+            new Function("updateNFTsMetadata(address,int64[],bytes)", ReturnTypes.INT);
+
+    private final UpdateDecoder decoder;
+
+    @Inject
+    public UpdateNFTsMetadataTranslator(final UpdateDecoder decoder) {
```
Comment (stoyanov-st) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/update/UpdateNFTsMetadataTranslator.java:41]:
```suggestion
    public UpdateNFTsMetadataTranslator(@NonNull final UpdateDecoder decoder) {
        requireNonNull(decoder);
```

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/token/UpdateTokenMetadataTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Tag;
+
+@Tag(SMART_CONTRACT)
+@DisplayName("metadataUpdateTests")
+@SuppressWarnings("java:S1192") // literals are duplicated for readability
+@HapiTestLifecycle
+public class UpdateTokenMetadataTest {
```
Comment (stoyanov-st) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/token/UpdateTokenMetadataTest.java:52]: You can add some negative cases here to validate that we return the correct responses.

**Review by stoyanov-st — DISMISSED** [review]
Thank you @MrValioBg!
Overall looks good with a small nit and suggestion for more tests.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/update/UpdateNFTsMetadataTranslator.java (hunk 2 above)
Comment (MrValioBg) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/update/UpdateNFTsMetadataTranslator.java]: Done. Great catch!

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/token/UpdateTokenMetadataTest.java (hunk 3 above)
Comment (MrValioBg) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/token/UpdateTokenMetadataTest.java:52]: Hey!
Could you please clarify what you refer to as "correct responses" and give an example of a test case?

We have a negative tests in `NumericValidationTests.failToUpdateNFTsMetadata` for the inputs.

Comment (MrValioBg) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/token/UpdateTokenMetadataTest.java:52]: Discussed offline, I will add some additional negative test cases for extra coverage, (currently not covered by TokenUpdateNftsSuite) and will explicitly check for the correct statuses in the childs

**Review by stoyanov-st — DISMISSED** [review]

**Review by stoyanov-st — APPROVED** [review]
LGTM

File: hedera-node/configuration/preprod/application.properties [hunk 4]
```diff
@@ -1,3 +1,4 @@
 contracts.chainId=298
 contracts.systemContract.updateCustomFees.enabled=true
 entities.unlimitedAutoAssociationsEnabled=true
+contracts.systemContract.updateNFTsMetadata.enabled=true
```
Comment (MrValioBg) [inline hedera-node/configuration/preprod/application.properties:4]: Should we also enable it for other environments, i.e. previewnet?

Comment (lukelee-sl) [reply hedera-node/configuration/preprod/application.properties:4]: I would not enable this until the metadata HIP has been implemented.

Comment (lukelee-sl) [reply hedera-node/configuration/preprod/application.properties:4]: I think I would wait until the metadata HIP has been implemented before turning this on

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Review by lukelee-sl — DISMISSED** [review]
LGTM.  tyvm

**Review by stoyanov-st — DISMISSED** [review]
LGTM

**Review by lukelee-sl — APPROVED** [review]
LGTM.  tyvm

**Review by Nana-EC — APPROVED** [review]

**Review by netopyr — DISMISSED** [review]
LGTM in regard to Services code

**Comment by stoyanov-st** [comment]
Closing this PR due to commit verification issues.
New PR with changes https://github.com/hashgraph/hedera-services/pull/15445

**Pull request closed without merging**